src/parser.js

```diff
@@ -34,99 +46,171 @@ class Parser {
    * @param {DOM} tree - DOM tree or Node
    * @param {Number} level - Current tree level
    */
-  * walk (tree, level) {
+  *walk(tree, level) {
     if (!tree) {
       return
     }
 
     for (let i = 0; i < tree.length; i++) {
       const node = tree[i]
 
-      this.pug += `\n${this.parseNode(node, level)}`
+      const newline = this.parseNode(node, level)
+      if (newline) {
+        this.pug += `\n${newline}`
+      }
 
       if (
         node.childNodes &&
         node.childNodes.length > 0 &&
-        !this.isUniqueTextNode(node)
+        !hasSingleTextNodeChild(node)
       ) {
-        yield * this.walk(node.childNodes, level + 1)
+        yield* this.walk(node.childNodes, level + 1)
       }
     }
   }
 
-  parseComment (node, indent) {
-    const comment = node.data.split('\n')
-
-    // Differentiate single line to multi-line comments
-    if (comment.length > 1) {
-      const multiLine = comment.map(line => `${indent}  ${line}`).join('\n')
-      return `${indent}//${multiLine}`
-    } else {
-      return `${indent}//${comment}`
-    }
-  }
-
-  parseNode (node, level) {
-    const indent = this.indent.repeat(level)
+  /*
+   * Returns a Pug node name with all attributes set in parentheses.
+   */
+  static getNodeWithAttributes(node) {
+    const { tagName, attrs } = node
+    const attributes = []
+    let pugNode = tagName
 
-    if (isDocumentTypeNode(node)) {
-      return `${indent}doctype html`
+    if (!attrs) {
+      return pugNode
     }
 
-    if (isTextNode(node)) {
-      return `${indent}| ${node.value}`
-    } else if (isCommentNode(node)) {
-      return this.parseComment(node, indent)
-    } else if (isElementNode(node)) {
-      let line = `${indent}${this.setAttributes(node)}`
+    // Add CSS selectors to pug node and append any element attributes to it
+    for (const attr of attrs) {
+      const { name, value } = attr
 
-      if (this.isUniqueTextNode(node)) {
-        line += ` ${node.childNodes[0].value}`
+      // Remove div tag if a selector is present (shorthand)
+      // e.g. div#form() -> #form()
+      const hasSelector = name === 'id' || name === 'class'
+      if (tagName === DIV_NODE && hasSelector) {
+        pugNode = pugNode.replace(DIV_NODE, '')
       }
 
-      return line
-    } else {
-      return node
-    }
-  }
-
-  setAttributes (node) {
-    const { tagName, attrs: attributes } = node
-    const attributeList = []
-    let pugNode = tagName
-
-    attributes.forEach(({ name, value }) => {
-      let shorten = false
-
       switch (name) {
```

Comment:
Switch inside of the `for` loop doesn't look particularly elegant, i'd just make a function out of it also to avoid `a += ${bar}` like things...

Reply:
I don't think there's anything inelegant about this, especially in a `for...of`.

Reply:
May be its a taste thing but in this switch and the `for` loop you are 1) mutating parent's scope prop and 2) mutating attributes from the same parent. So i'd extract whatever you have in the switch to return a value and then based on some condition make a call to either push to attributes or return a new pugNode. 
```
for (....) {
  const pugNode = getPugNode(name);
  const val = value.replace(/'/g, "\\'");

  if (val) {
    attributes.push(val ? `${name}='${val}'` : name)
  } 
}
...
return ...;
```

but i would also try to not use `for` and go with a reducer somehow as it seems to be a reducer in the end... Basically whatever it takes to escape parents scope props mutations. But again this is just an opinion =)

Reply:
@odincov Thanks for the explanation. This method is concatenating (or replacing) a string by looping over the attributes, so unfortunately your suggestion to put `const pugNode = getPugNode(name);` inside the `for` loop wouldn't work. I see what you're saying, but in building a string from an iterable list, is there any way to avoid parent scope mutations?